packages/grant-explorer/src/features/contributors/ViewContributionHistory.tsx

```diff
@@ -0,0 +1,395 @@
+import { useAccount, useEnsName } from "wagmi";
+import {
+  Client as AlloIndexerClient,
+  DetailedVote as Contribution,
+} from "allo-indexer-client";
+import { ChevronRightIcon } from "@heroicons/react/24/solid";
+import { useEffect, useState } from "react";
+import { useParams } from "react-router-dom";
+import { ethers } from "ethers";
+import { PayoutToken } from "../api/types";
+import { CHAINS } from "../api/utils";
+import Navbar from "../common/Navbar";
+import { ReactComponent as DonationHistoryBanner } from "../../assets/donnation-history-banner.svg";
+import { ChainId } from "../api/utils";
+import blockies from "ethereum-blockies";
+import CopyToClipboardButton from "../common/CopyToClipboardButton";
+import Footer from "common/src/components/Footer";
+import { payoutTokens } from "../api/utils";
+
+type ContributionHistoryState =
+  | { type: "loading" }
+  | {
+      type: "loaded";
+      data: { chainId: number; data: Contribution[] | never[] }[];
+    }
+  | { type: "error"; error: string };
+
+const useContributionHistory = (chainIds: number[], rawaddress: string) => {
+  const [state, setState] = useState<ContributionHistoryState>({
+    type: "loading",
+  });
+
+  useEffect(() => {
+    console.log("loading", chainIds, rawaddress);
+
+    if (!process.env.REACT_APP_ALLO_API_URL) {
+      throw new Error("REACT_APP_ALLO_API_URL is not set");
+    }
+
+    const histories: { chainId: number; data: Contribution[] | never[] }[] = [];
+    let allChainsError = true;
+
+    const fetchContributions = async () => {
+      const fetchPromises = chainIds.map((chainId: number) => {
+        if (!process.env.REACT_APP_ALLO_API_URL) {
+          throw new Error("REACT_APP_ALLO_API_URL is not set");
+        }
+
+        const client = new AlloIndexerClient(
+          fetch.bind(window),
+          process.env.REACT_APP_ALLO_API_URL,
+          chainId
+        );
+
+        let address = "";
+        try {
+          // ensure the address is a valid address
+          address = ethers.utils.getAddress(rawaddress.toLowerCase());
+        } catch (e) {
+          setState({
+            type: "error",
+            error: "bad address",
+          });
+        }
+
+        return client
+          .getContributionsByAddress(address)
+          .then((data) => {
+            const history = { chainId, data };
+            histories.push(history);
+            allChainsError = false; // If data is fetched successfully for any chain, set allChainsError to false
```

Comment:
setting a variable outside 3 levels of clousures worries me, another way you could do this more robustly is like this:

```
const chains = [1,2,3];
const myPromises: {error?: string, data: Contribution[] }[] = Promise.all(myArray.map((chainId) => {
     return getContributions(chainId, address).then(data => {data}).catch(error => {data: [], error});
}));

// then here we can check which one failed and which one was successful
```